sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -214,18 +251,62 @@
  */
 @Experimental
 public class KafkaIO {
+
+  /**
+   * Get a {@link Coder}. When a non-null coder is provided, use this coder. Otherwise,
+   * attempt to infer a {@link Coder} by extracting the type of the deserialized-class
+   * from the deserializer argument using the {@link Coder} registry.
+   */
+  @Nullable
+  public static <T> Coder<T> getOrInferCoder(CoderRegistry coderRegistry,
+                                             Class<? extends Deserializer<T>> deserializer,
+                                             Coder<T> coder) {
+    if (coder != null) {
+      return coder;
+    }
+
+    if (deserializer == null) {
+      LOG.warn("Could not extract deserializer type from {}", deserializer);
+      return null;
+    }
+
+    // Get Deserializer<T>
+    Type type = deserializer.getGenericInterfaces()[0];
```

Comment:
@peay, this method really needs some unit tests. Could you add them to KafkaIOTest.java?